chore(api-metrics): clean up exports

pichlermarc opened the pull request:
## Which problem is this PR solving?

`@opentelemetry/api-metrics` exports some classes and constants that are intended for internal use. In an effort to reduce the API surface before declaring metrics GA, this PR replaces the existing wildcard exports with explicit exports.

Related: https://github.com/open-telemetry/opentelemetry-js/issues/3158

## Short description of the changes
Replaces the existing wildcard exports with explicit exports and removes export for:
  - `NoopMeter`
  - `NoopMetric`
  - `NoopCounterMetric`
  - `NoopUpDownCounterMetric`
  - `NoopHistogramMetric`
  - `NoopObservableMetric`
  - `NoopObservableCounterMetric`
  - `NoopObservableGaugeMetric`
  - `NoopObservableUpDownCounterMetric`
  - `NOOP_COUNTER_METRIC`
  - `NOOP_HISTOGRAM_METRIC`
  - `NOOP_UP_DOWN_COUNTER_METRIC`
  - `NOOP_OBSERVABLE_COUNTER_METRIC`
  - `NOOP_OBSERVABLE_GAUGE_METRIC`
  - `NOOP_OBSERVABLE_UP_DOWN_COUNTER_METRIC`
  - `NoopMeterProvider`
  - `NOOP_METER_PROVIDER`

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- [x] Existing tests

## Checklist:

- [ ] Followed the style guidelines of this project
- [ ] Unit tests have been added
- [ ] Documentation has been updated

legendecas commented on experimental/packages/opentelemetry-api-metrics/src/index.ts:
```diff
@@ -14,13 +14,45 @@
-export * from './types/Metric';
-export * from './types/ObservableResult';
+export {
+  NOOP_METER,
```
Do we need to export noop instances? I didn't find noop instances are exported in `@opentelemetry/api`.

pichlermarc replied:
Yes, looks like it is used in the MeterProvider (returned for new meters when the provider is shutdown) and in the `@opentelemetry/node-sdk` package. I'll see if I can apply some refactoring to make it more similar to `@opentelemetry/api`.

Flarna replied:
Instead of exporting the NOOP instance itself we could export an API to get a noop meter. This would avoid that people rely on existence of such an instance or start to compare to it. it would be also possible to change the implementation to return a fresh Noop instead always the same in case it turns out that this is needed.

pichlermarc replied:
That's a good idea. I added this in [c8cba2e](https://github.com/open-telemetry/opentelemetry-js/pull/3198/commits/c8cba2e088ec71a3267e1f5547839206e604638c). :slightly_smiling_face:

legendecas commented on experimental/packages/opentelemetry-instrumentation/test/common/autoLoader.test.ts:
```diff
@@ -15,16 +15,28 @@
+
+class DummyMeterProvider implements MeterProvider {
+  getMeter(name: string, version?: string, options?: MeterOptions): Meter {
+    return NOOP_METER;
```
Can this method throw instead like the DummyTracerProvider above?

pichlermarc replied:
Yes that's possible. Did that in [d04eed3](https://github.com/open-telemetry/opentelemetry-js/pull/3198/commits/d04eed3ef85eefd698f693b04af22aa9329e5004), thanks! :slightly_smiling_face:. 

However, `NOOP_METER` is still used in the SDK's `MeterProvider`. I'm still looking into removing that, but the spec states that meter creation needs to return a valid [no-op implementation](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#meter-creation) when shutdown, while `TracerProvider` [does not](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracer-creation) and [does deliver a "normal" `Tracer`... ](https://github.com/open-telemetry/opentelemetry-js/blob/3db10568be4387ef024aa431ca1addef18c2441e/packages/opentelemetry-sdk-trace-base/src/BasicTracerProvider.ts#L93)

legendecas replied:
Cool, thank you for identifying this problem. I think the tracer provider can be tracked in a separate issue. And we can leave the `NOOP_METER` exported.

dyladan replied:
The spec is the same for tracer and we're not honoring that. The spec wording _if possible_ leaves this slightly open to interpretation but we should probably fix that as a bug in the tracer provider.

pichlermarc replied:
@legendecas Sounds good, thank you :slightly_smiling_face: I created an issue here https://github.com/open-telemetry/opentelemetry-js/issues/3211
@dyladan Ah, true. I missed that this is written in the "shutdown" section of the TracerProvider spec. :slightly_smiling_face:

legendecas commented on experimental/CHANGELOG.md:
```diff
@@ -15,6 +15,25 @@ All notable changes to experimental packages in this project will be documented
+    * `NOOP_OBSERVABLE_COUNTER_METRIC`
+    * `NOOP_OBSERVABLE_GAUGE_METRIC`
+    * `NOOP_OBSERVABLE_UP_DOWN_COUNTER_METRIC`
+    * `NoopMeterProvider`
```
Nit: would you mind sorting the list? i.e. move this item above the NOOP instances.

pichlermarc replied:
Oh just saw that @dyladan did that in [5581d5b](https://github.com/open-telemetry/opentelemetry-js/pull/3198/commits/5581d5b302af309094b529ee4acf104ce79784b7). Thanks for the feedback (and thx to Dan for sorting) :slightly_smiling_face:

pichlermarc commented:
Sorry for the force-push, something went wrong when I tried to merge main into my branch. Just removed the last commit and replaced it with a merge commit.

legendecas commented on experimental/packages/opentelemetry-api-metrics/src/NoopMeter.ts:
```diff
@@ -132,3 +132,10 @@ export const NOOP_UP_DOWN_COUNTER_METRIC = new NoopUpDownCounterMetric();
+/**
+ * Create a no-op Meter
+ */
+export function createNoopMeter(): Meter{
```
```suggestion
export function createNoopMeter(): Meter {
```

dyladan replied:
Wonder why lint didn't catch that
